src/crimson/os/seastore/async_cleaner.cc

```diff
@@ -1160,31 +1160,46 @@ SegmentCleaner::clean_space_ret SegmentCleaner::clean_space()
   // Backref-tree doesn't support tree-read during tree-updates with parallel
   // transactions.  So, concurrent transactions between trim and reclaim are
   // not allowed right now.
-  return extent_callback->with_transaction_weak(
-      "retrieve_from_backref_tree",
-      [this](auto &t) {
-    return backref_manager.get_mappings(
-      t,
-      reclaim_state->start_pos,
-      reclaim_state->end_pos
-    ).si_then([this, &t](auto pin_list) {
-      if (!pin_list.empty()) {
-	auto it = pin_list.begin();
-	auto &first_pin = *it;
-	if (first_pin->get_key() < reclaim_state->start_pos) {
-	  // BackrefManager::get_mappings may include a entry before
-	  // reclaim_state->start_pos, which is semantically inconsistent
-	  // with the requirements of the cleaner
-	  pin_list.erase(it);
-	}
-      }
-      return backref_manager.retrieve_backref_extents_in_range(
-        t,
-        reclaim_state->start_pos,
-        reclaim_state->end_pos
-      ).si_then([pin_list=std::move(pin_list)](auto extents) mutable {
-        return std::make_pair(std::move(extents), std::move(pin_list));
+  return seastar::do_with(
+    std::pair<std::vector<CachedExtentRef>, backref_pin_list_t>(),
+    [this](auto &weak_read_ret) {
+    return repeat_eagain([this, &weak_read_ret] {
```

Comment:
I don't see why the weak transaction needs eagain here.

> Since there won't be direct conflicts on root_t::backref_root, the backref retrieval procedule with weak transaction should be able to proceed. .
>
> Probably the only thing needed is to allow the weak transaction to tolerate the invalidation in the root-node?

I meant "tolerate" here by proceeding the weak transaction without retry, because there should be no one being able to modify the backref-tree at the same time, right?

Reply:
> I meant "tolerate" here by proceeding the weak transaction without retry, because there should be no one being able to modify the backref-tree at the same time, right?

The current main branch code is doing exactly this. But, this means that there are times in which backref retrieval transactions would proceed with an INVALID RootBlock, which breaks the invariant that transactions can't go on referencing an INVALID extent.

There are a lot of assertions that check on this invariant, so I submit this PR to fix this issue. I think transactions conflicting with others on RootBlock is a very rare case, so it should be ok to repeat the weak transaction in this case, am I right?

Reply:
I think the two directions are sementically different: tolerate that the root node can be invalid in the weak transaction, and tolerate that the weak transaction can retry.

They don't have the same impact from maintainence/debugging point of view.

> There are a lot of assertions in seastore that check on this invariant, so I submit this PR to fix this issue.

IIUC the root node read path is special that doesn't share much with other commen extents. Probably the PR https://github.com/ceph/ceph/pull/47749 further complicates the checks, but I haven't been able to check it thoroughly yet.

If it is complicate to do a reasonable work-around, probably "allow transactions accessing different fields of RootBlock to go on without invalidating each other" is a once-for-all solution.

Reply:
Oh, the PR https://github.com/ceph/ceph/pull/47974/files#r1062188256 proposes to change this transaction from weak to non-weak.

I think it is also reasonable if you are good with it.

Reply:
> Oh, the PR https://github.com/ceph/ceph/pull/47974/files#r1062188256 proposes to change this transaction from weak to non-weak.
> 
> I think it is also reasonable if you are good with it.

Yes, this is the first direction that I mentioned in https://github.com/ceph/ceph/pull/50191#issue-1592885775. PR https://github.com/ceph/ceph/pull/47974 is making multiple transactions modify the same backref extent possible, so making the backref retrieval transaction non-weak makes sense.

However, for the time being, backref retrieval transactions can only conflict with other transactions on the `RootBlock`, and non-weak transactions require all extents accessed be added to the transactions' `read_set` which is a relatively cpu-consuming operation, it doesn't seem to be worthy to make backref retrieval transactions non-weak.

> tolerate that the weak transaction can retry.

Actually, I think allowing weak transactions to retry when they conflict with other transactions on the `RootBlock` is a simple solution to the problem with nearly no overhead. What do you think?

Reply:
This seems to me is an synthetic issue now.

@zhscn 
> "CLEANER_MAIN" and "CLEANRE_COLD" might conflict on the backref extents. Such extents located in the reclaimed range are going to be rewritten by "CLEANER_MAIN" and "CLEANER_COLD" also needs these backref extents to retrieve live extents on the cold tier.

I don't see "CLEANER_MAIN" and "CLEANER_COLD" will modify the backref-tree except for the plain rewrite. So I think the issue is instead that the transaction should not invalidate the rewritten extent in cache, if the transaction doesn't mutate it's content.

@xxhdx1985126 
> However, for the time being, backref retrieval transactions can only conflict with other transactions on the RootBlock, and non-weak transactions require all extents accessed be added to the transactions' read_set which is a relatively cpu-consuming operation, it doesn't seem to be worth it to make backref retrieval transactions non-weak.

I agree with you that it doesn't worth it to make backref retrieval transactions non-weak from my above understanding.

> Actually, I think allowing weak transactions to retry when they conflict with other transactions on the RootBlock is a simple solution to the problem with nearly no overhead. What do you think?

IIUC the goal of this PR is to implement a workaround for https://github.com/ceph/ceph/pull/47749.

But we need to make a consensus about the final design, and add TODO to remember, IIUC:
* Allow transactions accessing different fields of RootBlock to go on without invalidating each other;
* Don't invalidate the rewritten extent in Cache, if the transaction doesn't modify the extent;

It should be fine to add simple workaround if we can agree on the final direction. Does it make sense?

Reply:
> It should be fine to add simple workaround if we can agree on the final direction. Does it make sense?

Yeah, totally agree:-)

Reply:
> * Keep this transaction weak and without retry;
> * Allow transactions accessing different fields of RootBlock to go on without invalidating each other;
> * Don't invalidate the rewritten extent in Cache, if the transaction doesn't modify the extent;

I think I can implement this in a following PR as soon as this weak, so that other PRs won't be blocked by this issue for long.

Reply:
> But we need to make a consensus about the final design, and add TODO to remember, IIUC:
> 
> * Keep this transaction weak and without retry;
> * Allow transactions accessing different fields of RootBlock to go on without invalidating each other;
> * Don't invalidate the rewritten extent in Cache, if the transaction doesn't modify the extent;
> 
> It should be fine to add simple workaround if we can agree on the final direction. Does it make sense?

@cyx1231st After a second thought, `Allow transactions accessing different fields of RootBlock to go on without invalidating each other` and `Don't invalidate the rewritten extent in Cache, if the transaction doesn't modify the extent` may not be enough to keep the `backref retrieval transaction` weak, because the parent node of the rewritten extent will always be modified, so there are always chances that `backref retrieval transaction` colide with `trim_dirty` on some backref extent after https://github.com/ceph/ceph/pull/47974#discussion_r1113865863 is merged.

Since we want to parallelise `trim_alloc` and `trim_dirty` eventually, I think maybe we should make the `backref retrieval transactions` non-weak in this PR. What do you think?

Reply:
Hmm, non-weak transaction makes sense to me for the correctness.

How we should parallelize the background transactions will still need to be evaluated with end-performance impacts, though.

Reply:
> * Keep this transaction weak and without retry;
> * Allow transactions accessing different fields of RootBlock to go on without invalidating each other;
> * Don't invalidate the rewritten extent in Cache, if the transaction doesn't modify the extent;

@cyx1231st I think we can keep backref retrieval transactions weak by adding a fourth one to the above requirements:

- cleaner transactions don't invalidate one another

This is feasible, the reason is as follows:

1. modifications that cleaner transactions can do are only the following two:
   a. rewriting extents
   b. updating parent nodes of the rewritten extents
2. different cleaners clean different tiers

So even if transactions from different cleaners are operating on the same extent, their modifications to that extent must be irrelevant.

What do you think?

Reply:
> cleaner transactions don't invalidate one another

This looks vague to me, do you mean that the cleaner transaction don't invalidate the weak-transaction of backref retrival, when they are running in parallel?

Then it seems to me the only thing needed is to let the weak-transaction to tolerate invalidations not only from root but also from backref-nodes.

As you previously said, the current main branch code is doing exactly this, except that some INVALID checks need to be relaxed for the weak transaction, right?

Reply:
> As you previously said, the current main branch code is doing exactly this, except that some INVALID checks need to be relaxed for the weak transaction, right?

Yes, but I still think we need to keep the invariant that no transaction can proceed referencing INVALID extents and breaking it may complicate future debugging.

However, I think we can put this difference aside for the moment, because, if `Transaction::add_to_read_set()` doesn't turn out to be the performance bottleneck, we can simply make backref retrieval transactions non-weak.

So I've just turned this PR's strategy to making backref retrieval transactions non-weak, please take a look again, thanks:-)

Reply:
@cyx1231st To avoid the unnecessary `Transaction::add_to_read_set()` overhead, I think we can limit the types of extents on which an transaction may conflict with other transactions, extents outside that set of extent types won't be added to Transactions' read_set, and in `Cache::invalidate_extent()`, we can assert that the extent being invalidated is in that extent types set. What do you think?

Reply:
Weak transaction is a special case in SeaStore, IIUC it is only used here in the no-test case (except for mnt).

Comparing making weak transaction tolerate the become-invalid extents, with not adding some extents to the read-set for all the transactions, I still think the previous one is more appropriate as a final solution.